src/compiler/checker.ts

```diff
@@ -9428,6 +9504,18 @@ namespace ts {
         }
 
         function checkApplicableSignature(node: CallLikeExpression, args: Expression[], signature: Signature, relation: Map<RelationComparisonResult>, excludeArgument: boolean[], reportErrors: boolean) {
+            const headMessage = Diagnostics.Argument_of_type_0_is_not_assignable_to_parameter_of_type_1;
+            if (signature.thisType && signature.thisType !== voidType && node.kind !== SyntaxKind.NewExpression) {
+                // If the source is not of the form `x.f`, then sourceType = voidType
+                // If the target is voidType, then the check is skipped -- anything is compatible.
+                // If the the expression is a new expression, then the check is skipped.
+                const calleeNode = node.kind === SyntaxKind.CallExpression && (<PropertyAccessExpression>(<CallExpression>node).expression).expression;

```

Comment:
This check isn't safe. You need to check if `(node as CallExpression).expression.kind` is a property access or element access, and only then can you extract the target expression. I think you need a `getThisArgumentOfCall` helper that does the extraction.